### PR TITLE
Add mypy

### LINKED ISSUE #431: add type hints to the project
### Description the feature
I think it would be nice to add type hints to the codebase, as it helps with documentation, catching bugs, and to maintain the whole project.

### Is your feature request related to a problem?
Nope

### Do you have any suggestions on how to add this feature in scanapi ?
Well, you just need to specify the type of every variable/parameter. I would be glad to do it myself, if you guys agree.

### PR DESCRIPTION
**Pull request by Stannislav**
## Description
This PR add static type checking to the project via `mpypy`.
- New dev dependencies: `mypy` + type stub packages
- New config file: `.mypy.ini`
- New command `make mypy`

This is an incremental configuration. This means that only module-level code is type checked, as well as all objects that already have type annotations. Everything else is skipped.

## Motivation behind this PR?
As described in #431 static type checking increases code quality, provides parameter documentation, and helps to identify bugs early.

## What type of change is this?
Adding a new dev-feature for code quality.

## Checklist
 

- [ ]  I have added a changelog entry / my PR does not need a new changelog entry. [Instructions](https://github.com/scanapi/scanapi/wiki/Changelog).
- [x] I have added/updated unit tests. [Instructions](https://github.com/scanapi/scanapi/wiki/Writing-Tests).
- [ ] New and existing unit tests pass locally with my changes. [Instructions](https://github.com/scanapi/scanapi/wiki/Run-ScanAPI-Locally#tests)
- [x] I have self-documented code my changes by adding docstring(s) and comment(s). [Instructions](https://github.com/scanapi/scanapi/wiki/First-Pull-Request#7-make-your-changes)
- [x] Current PR does not significantly decrease the code coverage and docstring coverage.
- [x] My code follows the style guidelines of this project.
- [x] I have run ScanAPI locally and manually tested my changes. [Instructions](https://github.com/scanapi/scanapi/wiki/Run-ScanAPI-Locally).

## Issue
Closes #431

### CODE REVIEW

**Comment by Stannislav** [comment]
It would probably make sense to add mypy to the CI. Need help for this one @camilamaia @Pradhvan.

**Comment by Stannislav** [comment]
There are some type issues in the `scanapi.config_loader` module, which alrady has type annotations. Currently investigating.

**Comment by Stannislav** [comment]
- Follow-up: fix `mypy` errors in `config_loader`: #457 
- Follow-up: add `mypy` to the CI: #458

File: Makefile
Comment (Pradhvan) [inline Makefile:16]: @Stannislav hey! 👋🏾  I think we can revert one commit back where `mypy` is no commented out. 

Once https://github.com/scanapi/scanapi/issues/457 is resolved. We can merge this PR just after that. What do you think ?

Also, you might need to add a changelog entry for this change.

**Review by Pradhvan — CHANGES_REQUESTED** [review]

File: Makefile
Comment (Stannislav) [reply Makefile:16]: Hey, thanks for the review! I thought this PR might be merged first, that's why I comment out for the CI to pass. But up to you, we can also wait for the other one :)

Comment (Pradhvan) [reply Makefile:16]: if #457 is not that much work we can wait for it to be merged and then merge this. That way we would not have to make an extra PR just to uncomment `mypy` and CI would be happy too. :)

Comment (Pradhvan) [reply Makefile:16]: @Stannislav On commenting out `mypy` we might have to create an extra PR just to uncomment. So I guess it would be best if we quickly patch up #457 and then we can merge this PR.

File: CHANGELOG.md
Comment (camilamaia) [inline CHANGELOG.md:9]: Since it is just a dev change which does not effect the final user, we don't need a changelog entry.

Here you can find when a new entry is needed or not: https://github.com/scanapi/scanapi/wiki/Changelog#what-warrants-a-changelog-entry

File: Makefile
Comment (camilamaia) [reply Makefile:16]: Or we can have the extra commit and I can squash everything in the end via github interface. It works fine for me too

**Review by camilamaia — COMMENTED** [review]
Thanks for the contribution! I can help with the CI tomorrow if you want.

File: Makefile
Comment (Stannislav) [reply Makefile:16]: Just submitted a new PR for fixing types in `config_loader`: #462

**Comment by Stannislav** [comment]
@camilamaia thanks for the review. I  removed the changelog entry in 11c2650

I also just submitted another PR for fixing the type annotations in `config_loader`: #462 With that a possible workflow could be
1. Review / merge #462 
2. Merge main into this PR
3. Uncomment `mypy` in `make check` in this PR
4. Review / merge this PR

File: Makefile
Comment (Pradhvan) [reply Makefile:16]: Since there is already a PR for #456, let's just wait and merge that. Then we can merge this and look into the documentation process of type annotations like how to start the whole thing, when should types be necessary for a PR and should they be included in the `Good First Issues` or `First Contribution` since folks who are starting out might not be familiar with it.

File: CHANGELOG.md
Comment (Stannislav) [reply CHANGELOG.md]: Fixed in 11c2650

**Comment by Pradhvan** [comment]
@Stannislav  Since this dependency PR #462  is merged. You can probably squash the commit into one and uncomment mypy now. I guess the CI should work fine now.

**Review by Pradhvan — APPROVED** [review]
LGTML 🚀

File: Makefile
Comment (camilamaia) [inline Makefile:12]: @Stannislav would you mind adding `mypy` to [.PHONY](https://github.com/scanapi/scanapi/blob/main/Makefile#L33)?

**Review by camilamaia — COMMENTED** [review]
Just one minor detail and this should be good to go 🚀  Thanks!!!

File: Makefile
Comment (Stannislav) [reply Makefile:12]: Ah yes, of course, totally missed it!

Done in 8363fed.

**Review by camilamaia — APPROVED** [review]

**Pull request merged**
